utils/oval/oval.go

```diff
@@ -0,0 +1,426 @@
+	ParseFilenameDist(line string) string
+	// OvalUri returns the url where the oval definitions are stored for a given
+	// distributions. See examples:
+	//   https://www.redhat.com/security/data/oval/
+	//   http://ftp.suse.com/pub/projects/security/oval/
+	OvalURI() string
+	// DistName returns the distribution name. Mostly used for debugging purposes
+	DistName() string
+	// IgnoredCriterions returns a list of strings that must be ignored when
+	// parsing the criterions.
+	// Oval parses parses all criterions by default trying to identify either
+	// package name and version or distribution version
+	IgnoredCriterions() []string
+	// SecToken returns a string that is compared with the value of reference.source
+	// in order to know if that is a security reference for, for example, using
+	// its url value
+	// Example return values: CVE, RHSA
+	SecToken() string
+	//Namespace stores the namespace that will be used in clair to store the vulnerabilities
+	Namespace() string
+}
+
+var (
+	log = capnslog.NewPackageLogger("github.com/coreos/clair", "utils/oval")
+)
+
+// FetchUpdate gets vulnerability updates from the OVAL definitions.
+func (f *OvalFetcher) FetchUpdate(datastore database.Datastore) (resp updater.FetcherResponse, err error) {
+	log.Info("fetching %s vulnerabilities", f.OsInfo.DistName())
+
+	if err != nil {
+		return resp, err
+	}
+
+	r, err := http.Get(f.OsInfo.OvalURI())
+
+	if err != nil {
+		log.Errorf("could not download %s's update list: %s", f.OsInfo.DistName(), err)
+		return resp, cerrors.ErrCouldNotDownload
+	}
+
+	var distList []string
+	scanner := bufio.NewScanner(r.Body)
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		filename := f.OsInfo.ParseFilenameDist(line)
+		if filename != "" {
+			distList = append(distList, filename)
+		}
+	}
+
+	for _, filename := range distList {
+		r, err := http.Get(filename)
+
+		if err != nil {
+			log.Errorf("could not download %s's update file: %s", f.OsInfo.DistName(), err)
+			return resp, cerrors.ErrCouldNotDownload
+		}
+
+		vs, err := f.ParseOval(r.Body)
+
+		if err != nil {
+			return resp, err
+		}
+
+		for _, v := range vs {
+			resp.Vulnerabilities = append(resp.Vulnerabilities, v)
+		}
+	}
+
+	// Set the flag if we found anything.
+	if len(distList) > 0 {
+		resp.FlagName = f.OsInfo.DistName() + "updater"
+		resp.FlagValue = distList[len(distList)-1]
+	} else {
+		log.Debug("no #{f.OsInfo.DistroName()}")
+	}
+
+	return resp, nil
+}
+
+// Clean deletes any allocated resources.
+func (f *OvalFetcher) Clean() {}
+
+// Parse criterions into an array of FeatureVersion for storing into the database
+func (f *OvalFetcher) ToFeatureVersions(possibilities [][]criterion) []database.FeatureVersion {
+	featureVersionParameters := make(map[string]database.FeatureVersion)
+
+	for _, criterions := range possibilities {
+		var (
+			featureVersion database.FeatureVersion
+			osVersion      string
+		)
+
+		for _, c := range criterions {
+			tmp_v := f.OsInfo.ParseOsVersion(c.Comment)
```

Comment:
before this line we could check to see if we have both a valid featureVersion and osVersion and break out of the loop